src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1231,137 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private static bool IsContextualModifier(SyntaxToken token)
+        {
+            switch (token.ContextualKind)
+            {
+                case SyntaxKind.PartialKeyword:
+                case SyntaxKind.AsyncKeyword:
+                    return true;
+            }
+
+            return false;
+        }
+
+        // Returns true if the current token is probably a modifier.
+        // To avoid further lookahead, caller is responsible to disambiguate some edge cases.
+        // For instance, we return true for both of these:
+        //
+        //    partial partial<T>()
+        //    partial partail<T> partial()
+        //
+        // While, in fact, 'partial' is only a modifier on the second method.
+        private bool IsPossibleModifier()
+        {
+            if (!IsAnyModifier(this.CurrentToken))
+            {
+                return false;
+            }
+
+            if (IsContextualModifier(this.CurrentToken))
+            {
+                var tk = this.PeekToken(1);
+                return tk.Kind == SyntaxKind.IdentifierToken ||
+                    IsPredefinedType(tk.Kind) ||
+                    IsTypeDeclarationStart(tk.Kind) || 
+                    IsAnyModifier(tk);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        ///  Possible results from scanning for a partial member or type.
+        /// </summary>
+        private enum ScanPartialFlags
+        {
+            /// <summary>
+            /// Definitly a partial type.
+            /// </summary>
+            PartialType,
+            /// <summary>
+            /// Definitly a partial member.
+            /// </summary>
+            PartialMember,
+            /// <summary>
+            /// Treat as modifier for better diagnostics to be reported during binding.
+            /// </summary>
+            TreatAsModifier,
+            /// <summary>
+            /// Not a modifier.
+            /// </summary>
+            NotModifier
+        }
+
+        private ScanPartialFlags ScanPartialTypeOrMember()
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
+
+            var point = this.GetResetPoint();
+
+            try
+            {
+                EatToken(); // partial
+
+                SyntaxToken lastModOpt = null;
+
+                // Skip over additional modifiers
+                while (IsPossibleModifier())
```

Comment:
>while [](start = 16, length = 5)

I don't believe the intent is to permit `partial public class C {}`. The language change that has been approved is to permit `modifiers partial ref classOrStruct ...` and also `modifiers ref partial classOrStruct ...`

For that change, I believe a much simpler implementation is possible.

Reply:
That was not my understanding. I think we should check with LDM again.

Reply:
@gafter I think that's incorrect. See the championed issue: https://github.com/dotnet/csharplang/issues/946

Reply:
limiting to just "ref partial" or "partial ref" would also bum me out :)   I would much rather prefer that the modifiers be allowed in any order.  the old "partial must be last" constraint really came out of the old parser and trying to keep cost low by not having the parser have to worry about so many subtle cases.  But, from teh language perspective, i would far prefer it's just a modifier, allowed in any order, and the impl is smart enough to understand when it's a modifier vs when it's not.

Reply:
Well, we do not just change the language by accepting a PR. We need LDM signoff. The championed issue says "It seems likely that various other keywords could be used to disambiguate these contextual modifiers and allow us to relax the constraints on where `partial` and `ref` can appear in the modifier list.", and I agree that it seems likely. But that's not a definitive spec. The LDM's pronouncement on this is at: https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-12-04.md#this-ref-vs-ref-this and doesn't mention any generalizations. We don't have a spec, and I cannot review an implementation of a language feature for which we have no spec.

Reply:
I believe the LDM notes you linked is regarding parameter modifiers (https://github.com/dotnet/roslyn/pull/23643). Looks like LDM decision on this feature is not published yet (or it may never made).

Reply:
From discussion earlier today, LDM confirmed the desire to fully relax the order of modifiers, and treat "ref" and "partial" as modifiers when there are only modifiers between them and "class/struct".

Reply:
Right, that is what is implemented so far. The complexity here is to give "duplicated modifiers" for something like `partial partial void` etc instead of a couple of syntax errors.

Reply:
Note: There is a followup change to migrate the same diag clarity for `async` (https://github.com/alrz/roslyn/commit/6a4d586859559d7a949aef8b64b518f4593f4e0b).
If we still want this for 15.7, would you prefer to also take that followup for 15.7?